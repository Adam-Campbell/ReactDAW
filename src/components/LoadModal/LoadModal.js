import React from 'react';
import PropTypes from 'prop-types';

const LoadTrackModal = props => (
    <>
        <button
            className="button main-color"
            onClick={props.handleClose}
        >X</button>
        <div className="load-modal__container">
            
            <ul className="modal__list">
            {props.savedTracks.lenght ? props.savedTracks.map((trackName, index) => (
                <li
                    className="modal__list-item"
                    key={index}
                    onClick={(e) => props.handleLoad(trackName)}
                >{trackName}</li>
            )) : (
                <p className="modal__text">Sorry, there are no tracks available to load.</p>
            )}
            </ul>
        </div>
    </>
);

LoadTrackModal.propTypes = {
    handleClose: PropTypes.func.isRequired,
    savedTracks: PropTypes.arrayOf(PropTypes.string).isRequired,
    handleLoad: PropTypes.func.isRequired
};

export default LoadTrackModal;