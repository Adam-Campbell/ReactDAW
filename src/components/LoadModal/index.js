import React, { Component } from 'react';
import PropTypes from 'prop-types';
import * as ActionCreators from '../../actions';
import { connect } from 'react-redux';
import LoadModal from './LoadModal';

export const getSavedTracks = () => {
    const savedTracks = [];
    const len = window.localStorage.length;
    for (let i = 0; i < len; i++) {
        const key = window.localStorage.key(i);
        if (key.includes('[track_name]')) {
            const trimmedName = key.replace(/\[track_name\] /, '');
            savedTracks.push(trimmedName);
        }
    }
    return savedTracks;
}

export class LoadModalContainer extends Component {

    handleLoad = trackName => {
        const namespacedTrackName = `[track_name] ${trackName}`;
        this.props.loadState(namespacedTrackName);
        this.props.closeModal();
    }

    render() {
        const savedTracks = getSavedTracks();
        return (
            <LoadModal 
                handleClose={this.props.closeModal}
                handleLoad={this.handleLoad}
                savedTracks={savedTracks}
            />
        );
    }
}


// class LoadModalContainer extends Component {


//     handleLoad = (trackName) => {
//         const namespacedTrackName = `[track_name] ${trackName}`;
//         this.props.loadState(namespacedTrackName);
//         this.props.closeModal();
//     }

//     render() {
//         const savedTracks = this.getSavedTracks();
//         return <LoadModal 
//             handleClose={this.props.closeModal}
//             savedTracks={savedTracks}
//             handleLoad={this.handleLoad}
//         />
//     }
// }

export default connect(
    undefined, 
    {
        loadState: ActionCreators.loadState,
        closeModal: ActionCreators.closeModal
    }
)(LoadModalContainer);