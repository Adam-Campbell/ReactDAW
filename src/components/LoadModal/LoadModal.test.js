import React from 'react';
import { shallow } from 'enzyme';
import LoadModal from './LoadModal';

describe('LoadModal', () => {

    test('renders correctly', () => {
        const wrapper = shallow(
            <LoadModal 
                handleClose={jest.fn()}
                handleLoad={jest.fn()}
                savedTracks={['Track One', 'Track Two']}
            />
        ); 
        expect(wrapper).toMatchSnapshot();
    });

    test('renders a list item for each track in props.savedTracks', () => {
        const wrapper = shallow(
            <LoadModal 
                handleClose={jest.fn()}
                handleLoad={jest.fn()}
                savedTracks={['Track One', 'Track Two']}
            />
        ); 
        expect(wrapper.find('.modal__list-item')).toHaveLength(2);
        expect(wrapper.find('.modal__list-item').at(0).text()).toBe('Track One');
        expect(wrapper.find('.modal__list-item').at(1).text()).toBe('Track Two');
    });

    test('handleLoad is called with the appropriate argument depending on which list item is clicked on', () => {
        const mockedHandleLoad = jest.fn();
        const wrapper = shallow(
            <LoadModal 
                handleClose={jest.fn()}
                handleLoad={mockedHandleLoad}
                savedTracks={['Track One', 'Track Two']}
            />
        ); 
        wrapper.find('.modal__list-item').at(0).simulate('click');
        expect(mockedHandleLoad).toHaveBeenCalledTimes(1);
        expect(mockedHandleLoad.mock.calls[0][0]).toBe('Track One');
        wrapper.find('.modal__list-item').at(1).simulate('click');
        expect(mockedHandleLoad).toHaveBeenCalledTimes(2);
        expect(mockedHandleLoad.mock.calls[1][0]).toBe('Track Two');
    });

    test('handleClose is called when the appropriate action is taken', () => {
        const mockedHandleClose = jest.fn();
        const wrapper = shallow(
            <LoadModal 
                handleClose={mockedHandleClose}
                handleLoad={jest.fn()}
                savedTracks={['Track One', 'Track Two']}
            />
        ); 
        wrapper.find('.button').first().simulate('click');
        expect(mockedHandleClose).toHaveBeenCalledTimes(1);
    });

});