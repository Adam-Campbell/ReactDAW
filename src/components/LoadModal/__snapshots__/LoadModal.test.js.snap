// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadModal renders correctly 1`] = `
<Fragment>
  <button
    className="button pink"
    onClick={[MockFunction]}
  >
    Close
  </button>
  <ul
    className="modal__list"
  >
    <li
      className="modal__list-item"
      key="0"
      onClick={[Function]}
    >
      Track One
    </li>
    <li
      className="modal__list-item"
      key="1"
      onClick={[Function]}
    >
      Track Two
    </li>
  </ul>
</Fragment>
`;
